### PR TITLE
CompatHelper version 3: major rewrite and overhaul of the entire package (i.e. merge the `v3-dev` branch into the `master` branch)

### PR DESCRIPTION
**Pull request by DilumAluthge**

### CODE REVIEW

**Comment by DilumAluthge** [comment]
bors try

**Comment by DilumAluthge** [comment]
bors try

**Comment by DilumAluthge** [comment]
@fchorney @mattBrzezinski I think we should figure out these integration test failures before merging.

It looks like it's just that we're hitting some rate limits during the integration tests? If that's the case, we should do one or both of:
1. Wrap the relevant tests in a bounded retry loop, maybe with an exponential backoff.
2. Insert some sleep statements between tests

**Comment by fchorney** [comment]
weird that we haven't his these rate limits before during testing. Did GitHub change them? or is there some other reason we are hitting them?

**Comment by fchorney** [comment]
Hmm it also seems like it's the `master 3` test every time. Maybe something is wonky with that test specifically

**Comment by mattBrzezinski** [comment]
Looking at the [GitHub Documentation](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#secondary-rate-limits)

> additional rate limits may apply to some actions when using the API. For example, using the API to rapidly create content, poll aggressively instead of using webhooks, make multiple concurrent requests, or repeatedly request data that is computationally expensive may result in secondary rate limiting.

There is no getting around these limits. They do not interfere with the normal API. We need to retry / sleep here. Personally I think doing a `sleep` is a bit better as we're not spamming GitHub repeatedly. I'll make a PR for these changes.

**Comment by mattBrzezinski** [comment]
And to follow up [Best Practices for Integrators](https://docs.github.com/en/rest/guides/best-practices-for-integrators#dealing-with-secondary-rate-limits) says to wait one second between requests here. If you're still hitting it, do a retry using the `Retry-After` header from the response.

**Comment by DilumAluthge** [comment]
Anyway, once you've fixed the tests, this is good to go from my point of view.

I would say do the following steps:
1. Fix whatever needs fixing.
2. Run `bors try` and make sure it passes.
3. After you have a successful `bors try` run, convert this PR from "draft" to "ready to review".
4. Do `bors merge`.
5. Once the PR has been merged, you can register a new release with JuliaRegistrator.
6. Wait for the General PR to be merged.
7. Make an announcement on Discourse.

**Comment by mattBrzezinski** [comment]
bors try

**Review by mattBrzezinski — APPROVED** [review]

File: .github/workflows/format_pr.yml
Comment (DilumAluthge) [reply .github/workflows/format_pr.yml:5]: I would first merge v3-dev into master. Then make a quick PR to master to fix up things like this.

**Pull request merged**
